### PR TITLE
Implement instance norm in NNX

### PR DESCRIPTION
**Pull request by mattbahr**
Relates to #4623 - Brings the Instance Norm implementation into NNX from Linen.

Created this PR at the request of @vfdev-5 to merge the instance norm implementation separately from spectral norm.

### CODE REVIEW

**Comment by mattbahr** [comment]
@vfdev-5 I'm having issues getting my dev container working again, so I haven't been able to rerun the tests or update the docstring output. In any case, I'll strip instance norm out of #4623 but I think this is all I'll have time to do at the moment. Feel free to take over and update however you like, or to create your own PR off my commits! Thanks!

File: flax/nnx/nn/normalization.py [hunk 1]
```diff
@@ -832,4 +832,155 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+
+class InstanceNorm(Module):
+  """Instance normalization (https://arxiv.org/abs/1607.08022v3).
+  InstanceNorm normalizes the activations of the layer for each channel (rather
+  than across all channels like Layer Normalization), and for each given example
+  in a batch independently (rather than across an entire batch like Batch
+  Normalization). i.e. applies a transformation that maintains the mean activation
+  within each channel within each example close to 0 and the activation standard
+  deviation close to 1.
+  .. note::
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]:
```suggestion
  deviation close to 1.

  .. note::
```

File: flax/nnx/nn/normalization.py [hunk 2]
```diff
@@ -832,4 +832,155 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+    This normalization operation is identical to LayerNorm and GroupNorm; the
+    difference is simply which axes are reduced and the shape of the feature axes
+    (i.e. the shape of the learnable scale and bias parameters).
+  Example usage::
+    >>> from flax import nnx
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]:
```suggestion
    (i.e. the shape of the learnable scale and bias parameters).

  Example usage::

    >>> from flax import nnx
```

File: flax/nnx/nn/normalization.py [hunk 3]
```diff
@@ -832,4 +832,155 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+    })
+    >>> y = layer(x)
+    >>> # having a channel_axis of -1 in InstanceNorm is identical to reducing all non-batch,
+    >>> # non-channel axes and using the feature_axes as the feature_axes in LayerNorm
+    >>> y2 = nnx.LayerNorm(5, reduction_axes=[1, 2], feature_axes=-1, rngs=nnx.Rngs(0))(x)
+    >>> np.testing.assert_allclose(y, y2, atol=1e-7)
+    >>> y3 = nnx.GroupNorm(5, num_groups=x.shape[-1], rngs=nnx.Rngs(0))(x)
+    >>> np.testing.assert_allclose(y, y3, atol=1e-7)
+  Args:
+    num_features: the number of input features/channels.
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]:
```suggestion
    >>> np.testing.assert_allclose(y, y3, atol=1e-7)

  Args:
    num_features: the number of input features/channels.
```

File: flax/nnx/nn/normalization.py [hunk 4]
```diff
@@ -832,4 +832,155 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
...
+
+  def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
+    """Applies instance normalization on the input.
+    Args:
+      x: the inputs
+      mask: Binary array of shape broadcastable to ``inputs`` tensor, indicating
+        the positions for which the mean and variance should be computed.
+    Returns:
+      Normalized inputs (the same shape as inputs).
+    """
```
Comment (vfdev-5) [inline flax/nnx/nn/normalization.py]:
```suggestion
    """Applies instance normalization on the input.

    Args:
      x: the inputs
      mask: Binary array of shape broadcastable to ``inputs`` array, indicating
        the positions for which the mean and variance should be computed.

    Returns:
      Normalized inputs (the same shape as inputs).
    """
```

**Comment by vfdev-5** [comment]
Thanks @mattbahr for this PR on InstanceNorm, I'll make change directly here.
Thanks again for your contributions!

**Comment by cgarciae** [comment]
Triggered the internal review
